synapse/lib/persist.py

```diff
@@ -0,0 +1,114 @@
+        ``tqueue`` attribute which is a ``synapse.lib.queue.Queue`` object,
+        and a dictionary of configuration data.
+
+    Returns:
+        None
+    '''
+    qtypes[name] = func
+
+def getQueues():
+    '''
+    Get a list of registered queue names and their fully qualified paths.
+    '''
+    ret = []
+    for alias, ctor in qtypes.items():
+        ret.append((alias, '.'.join([ctor.__module__, ctor.__qualname__])))
+    return ret
+
+def run(core, conf):
+    func = qtypes.get(conf.get('type'))
+    func(core, conf)
+
+def cryoCellQueue(core, conf):
+    '''
+    Consume task events from the Cortex.tqueue, and place them into CryoTanks
+    managed by a CryoCell.
+
+    Args:
+        core: Cortex to consume events from.
+        conf (dict): A config dictionary. The following keys are used:
+          - url: The Cryocell to connect too via Telepath.
+          - size: The number of items to pull from the queue at once. If not
+          provided, the default size of 1000 will be used.
+
+    Notes:
+        Task events consumed by the Queue are bucketed by their task name, and
+        only the task dictionaries are placed into the CryoTank. Tasks are put
+        into the tanks by their name.
+        For Cortex configuration purposes, this is named ``cryo:cell``.
+
+    '''
+    url = conf.get('url')
+    sz = conf.get('size', 1000)
+    with s_telepath.openurl(url) as cryocell:
+        for items in core.tqueue.slices(sz):
+            d = collections.defaultdict(list)
+            for ttype, rec in items:
+                d[ttype].append(rec)
+
+            for ttype, recs in d.items():
+                try:
+                    cryocell.puts(ttype, recs)
+                except Exception as e:
+                    logger.exception('Failed to put items into tank for [%s]', ttype)
+                    raise
+            core.fire('persist:task:cryo:cell', len=len(items))
+
+def cryoTankQueue(core, conf):
+    '''
+        Consume task events from the Cortex.tqueue, and put them directly into
+        a Cryotank.
+
+        Args:
+            core: Cortex to consume events from.
+            conf (dict): A config dictionary. The following keys are used:
+              - url: The CryoTank to connect to via Telepath.
+              - size: The number of items to pull from the queue at once. If not
+              provided, the default size of 1000 will be used.
+
+        Notes:
+            This puts tasks into the CryoTanks as a tufo of
+            ``('someTaskName', {..task data..})``.
+            For Cortex configuration purposes, this is named ``cryo:tank``.
+
+        '''
+    url = conf.get('url')
+    sz = conf.get('size', 1000)
+    with s_telepath.openurl(url) as cryotank:
+        for items in core.tqueue.slices(sz):
+            try:
+                cryotank.puts(items)
+            except Exception as e:
+                logger.exception('Failed to put items into cryotank')
```

Comment:
Do we want to put items back into `core.tqueue()` prior to raising? That way no tasks are dropped on the floor because of a possibly remote exception.

Reply:
Same comment as the other one